File: packages/fess_pos_engine_ts/src/testcases.ts
/**
 * Definition test cases (docs/04 §9, B4.12): recorded scenarios of context + answer steps with
 * expected visibility, required flags, computed values, options, props, labels and validation errors.
 * Publishing runs every test case of the family; any failure blocks the publish.
 *
 * Answers are built the way the module builds a submission: visible fields only (hidden ⇒ absent),
 * computed fields from the resolver (`computed: true`), prefilled fields from the context.
 */
import type { JsonObject, JsonValue } from "./json.ts";
import { deepEqual, isPlainObject } from "./json.ts";
import { compileForm, resolveForm } from "./resolver.ts";
import type { ResolveContext, ResolveLists, ResolvedField, ResolvedForm } from "./resolver.ts";
import { validateAnswers } from "./validator.ts";
import type { AnswerEntry } from "./validator.ts";
import { hasValue } from "./definitions/catalogue.ts";
import type { FormDefinition } from "./definitions/types.ts";
import type { ValidationError } from "./validator.ts";

export interface TestExpectation {
  visible?: Record<string, boolean>;
  required?: Record<string, boolean>;
  read_only?: Record<string, boolean>;
  /** Effective values (computed or answered) by field path. */
  values?: Record<string, JsonValue>;
  options?: Record<string, string[]>;
  /** Subset match on resolved props. */
  props?: Record<string, JsonObject>;
  labels?: Record<string, string>;
  /** Exact set of validation errors (field_key + code). */
  errors?: { field_key: string; code: string }[];
  valid?: boolean;
}

export interface TestStep {
  set?: Record<string, JsonValue>;
  unset?: string[];
  expect?: TestExpectation;
}

export interface TestCase {
  name: string;
  context?: ResolveContext;
  lists?: ResolveLists;
  steps?: TestStep[];
  expect?: TestExpectation;
}

export interface TestFailure {
  readonly step: number | "final";
  readonly check: string;
  readonly path: string;
  readonly expected: JsonValue;
  readonly actual: JsonValue;
}

export interface TestCaseResult {
  readonly name: string;
  readonly passed: boolean;
  readonly failures: TestFailure[];
}

function lookup(resolved: ResolvedForm, path: string): ResolvedField | undefined {
  const m = /^([a-z][a-z0-9_]*)\[(\d+)\]\.([a-z][a-z0-9_]*)$/.exec(path);
  if (!m) return resolved.fields[path];
  const group = resolved.fields[m[1] as string];
  return group?.items?.[Number(m[2])]?.fields[m[3] as string];
}

/** Build a submission-style answers map from raw state, as the module would. */
export function buildAnswers(form: FormDefinition, resolved: ResolvedForm, state: Record<string, JsonValue>): Record<string, AnswerEntry> {
  const compiled = compileForm(form);
  const out: Record<string, AnswerEntry> = {};
  for (const cf of compiled.fields) {
    const key = cf.def.key;
    const rf = resolved.fields[key];
    if (!rf || !rf.visible || !hasValue(cf.spec)) continue;
    if (rf.computed) {
      if (rf.value !== null) out[key] = { v: rf.value, computed: true };
      continue;
    }
    if (cf.spec.type === "prefilled") {
      if (rf.value !== null) out[key] = { v: rf.value, prefilled: true };
      continue;
    }
    if (cf.spec.type === "repeatable_group" && Array.isArray(state[key])) {
      const items = (rf.items ?? []).map((item) => {
        const entry: Record<string, JsonValue> = {};
        for (const [ck, kf] of Object.entries(item.fields)) {
          if (!kf.visible || kf.type === "group" || kf.type === "info" || kf.type === "callout" || kf.type === "divider" || kf.type === "image") continue;
          if (kf.computed) {
            if (kf.value !== null) entry[ck] = { v: kf.value, computed: true };
          } else if (kf.type === "prefilled") {
            if (kf.value !== null) entry[ck] = { v: kf.value, prefilled: true };
          } else if (kf.value !== null) {
            entry[ck] = { v: kf.value };
          }
        }
        return entry as JsonValue;
      });
      out[key] = { v: items };
      continue;
    }
    const v = state[key];
    if (v !== undefined) out[key] = { v };
  }
  return out;
}

function check(exp: TestExpectation, resolved: ResolvedForm, errors: readonly ValidationError[], step: number | "final", failures: TestFailure[]): void {
  const fail = (check: string, path: string, expected: JsonValue, actual: JsonValue): void => {
    failures.push({ step, check, path, expected, actual });
  };
  const each = <T extends JsonValue>(m: Record<string, T> | undefined, name: string, get: (rf: ResolvedField) => JsonValue, eq: (a: T, b: JsonValue) => boolean = (a, b) => deepEqual(a, b)): void => {
    for (const [path, expected] of Object.entries(m ?? {})) {
      const rf = lookup(resolved, path);
      const actual = rf ? get(rf) : null;
      if (!rf || !eq(expected, actual)) fail(name, path, expected, rf ? actual : "<no such field>");
    }
  };
  each(exp.visible, "visible", (rf) => rf.visible);
  each(exp.required, "required", (rf) => rf.required);
  each(exp.read_only, "read_only", (rf) => rf.read_only);
  each(exp.values, "value", (rf) => rf.value);
  each(exp.labels, "label", (rf) => rf.label ?? null);
  each(exp.options, "options", (rf) => (rf.options ?? []).map((o) => o.value));
  each(exp.props, "props", (rf) => rf.props, (expected, actual) => {
    if (!isPlainObject(actual)) return false;
    return Object.entries(expected).every(([k, v]) => deepEqual(v, (actual as JsonObject)[k] ?? null));
  });
  if (exp.errors) {
    const want = exp.errors.map((e) => `${e.field_key}|${e.code}`).sort();
    const got = errors.map((e) => `${e.field_key}|${e.code}`).sort();
    if (!deepEqual(want, got)) fail("errors", "", want, got);
  }
  if (exp.valid !== undefined && exp.valid !== (errors.length === 0)) fail("valid", "", exp.valid, errors.length === 0);
}

export function runTestCase(form: FormDefinition, tc: TestCase): TestCaseResult {
  const state: Record<string, JsonValue> = {};
  const failures: TestFailure[] = [];
  const context = tc.context ?? {};
  const lists = tc.lists ?? {};
  const evaluateState = (): { resolved: ResolvedForm; errors: ValidationError[] } => {
    const resolved = resolveForm(form, context, state, lists);
    const answers = buildAnswers(form, resolved, state);
    const res = validateAnswers(form, answers, context, lists);
    return { resolved: res.resolved ?? resolved, errors: res.errors };
  };
  try {
    (tc.steps ?? []).forEach((step, i) => {
      for (const [k, v] of Object.entries(step.set ?? {})) state[k] = v;
      for (const k of step.unset ?? []) delete state[k];
      if (step.expect) {
        const { resolved, errors } = evaluateState();
        check(step.expect, resolved, errors, i, failures);
      }
    });
    if (tc.expect) {
      const { resolved, errors } = evaluateState();
      check(tc.expect, resolved, errors, "final", failures);
    }
  } catch (e) {
    failures.push({ step: "final", check: "exception", path: "", expected: null, actual: e instanceof Error ? e.message : String(e) });
  }
  return { name: tc.name, passed: failures.length === 0, failures };
}

export function runTestCases(form: FormDefinition, cases: readonly TestCase[]): { passed: boolean; results: TestCaseResult[] } {
  const results = cases.map((c) => runTestCase(form, c));
  return { passed: results.every((r) => r.passed), results };
}

File: packages/fess_pos_engine/lib/src/forms/testcases.dart
/// Definition test cases (docs/04 §9, B4.12), the Dart twin of
/// `src/testcases.ts`: recorded scenarios of context and answer steps with
/// the visibility, required flags, values, options, props, labels and
/// validation errors they must produce. Publishing runs them in the
/// TypeScript engine; the same fixtures run here, so both engines agree on
/// what a form does.
///
/// Answers are built the way the module builds a submission: visible
/// fields only (hidden ⇒ absent), computed fields from the resolver
/// (`computed: true`), prefilled fields from the context.
library;

import 'package:fess_pos_engine/src/forms/model.dart';
import 'package:fess_pos_engine/src/forms/resolver.dart';
import 'package:fess_pos_engine/src/forms/validator.dart';
import 'package:fess_pos_engine/src/json.dart';
import 'package:meta/meta.dart';

/// One expectation a test case didn't meet.
@immutable
class TestFailure {
  const TestFailure(
    this.step,
    this.check,
    this.path,
    this.expected,
    this.actual,
  );

  /// The step's index, or `'final'` for the case's own `expect`.
  final Object step;
  final String check;
  final String path;
  final Object? expected;
  final Object? actual;

  Map<String, Object?> toJson() => {
    'step': step,
    'check': check,
    'path': path,
    'expected': expected,
    'actual': actual,
  };

  @override
  String toString() =>
      'step $step: $check $path expected $expected, got $actual';
}

@immutable
class TestCaseResult {
  const TestCaseResult(this.name, this.failures);

  final String name;
  final List<TestFailure> failures;

  bool get passed => failures.isEmpty;
}

final RegExp _itemPath = RegExp(
  r'^([a-z][a-z0-9_]*)\[(\d+)\]\.([a-z][a-z0-9_]*)$',
);

/// A field by path: its key, or `group[i].key` inside a repeatable group.
ResolvedField? _lookup(ResolvedForm resolved, String path) {
  final m = _itemPath.firstMatch(path);
  if (m == null) return resolved.fields[path];
  final items = resolved.fields[m[1]]?.items;
  final index = int.parse(m[2]!);
  if (items == null || index >= items.length) return null;
  return items[index].fields[m[3]];
}

const Set<String> _noAnswer = {'group', 'info', 'callout', 'divider', 'image'};

Map<String, Object?> _entry(ResolvedField f) => f.computed
    ? {'v': f.value, 'computed': true}
    : f.type == 'prefilled'
    ? {'v': f.value, 'prefilled': true}
    : {'v': f.value};

/// A submission-style answers map (`{key: {v, …}}`) from the raw [state],
/// as the module builds it.
Map<String, Object?> buildAnswers(
  Map<String, Object?> form,
  ResolvedForm resolved,
  Map<String, Object?> state,
) {
  final compiled = compileForm(form);
  final out = <String, Object?>{};
  for (final cf in compiled.fields) {
    final key = cf.key;
    final rf = resolved.fields[key];
    if (rf == null || !rf.visible || !cf.spec.hasValue) continue;
    if (rf.computed || cf.spec.type == 'prefilled') {
      if (rf.value != null) out[key] = _entry(rf);
      continue;
    }
    if (cf.spec.type == 'repeatable_group' && state[key] is List<Object?>) {
      out[key] = {
        'v': [
          for (final item in rf.items ?? const <ResolvedItem>[])
            {
              for (final e in item.fields.entries)
                if (e.value.visible &&
                    !_noAnswer.contains(e.value.type) &&
                    e.value.value != null)
                  e.key: _entry(e.value),
            },
        ],
      };
      continue;
    }
    if (state.containsKey(key)) out[key] = {'v': state[key]};
  }
  return out;
}

void _check(
  Map<String, Object?> exp,
  ResolvedForm resolved,
  List<ValidationError> errors,
  Object step,
  List<TestFailure> failures,
) {
  void fail(String check, String path, Object? expected, Object? actual) =>
      failures.add(TestFailure(step, check, path, expected, actual));

  void each(
    String property,
    String check,
    Object? Function(ResolvedField rf) get, [
    bool Function(Object? expected, Object? actual) eq = deepEqual,
  ]) {
    final m = exp[property];
    if (m is! Map<String, Object?>) return;
    for (final e in m.entries) {
      final rf = _lookup(resolved, e.key);
      final actual = rf == null ? null : get(rf);
      if (rf == null || !eq(e.value, actual)) {
        fail(check, e.key, e.value, rf == null ? '<no such field>' : actual);
      }
    }
  }

  each('visible', 'visible', (rf) => rf.visible);
  each('required', 'required', (rf) => rf.required);
  each('read_only', 'read_only', (rf) => rf.readOnly);
  each('values', 'value', (rf) => rf.value);
  each('labels', 'label', (rf) => rf.label);
  each(
    'options',
    'options',
    (rf) => [
      for (final o in rf.options ?? const <OptionDef>[]) o.value,
    ],
  );
  each(
    'props',
    'props',
    (rf) => rf.props,
    (expected, actual) =>
        expected is Map<String, Object?> &&
        actual is Map<String, Object?> &&
        expected.entries.every((e) => deepEqual(e.value, actual[e.key])),
  );
  final wantErrors = exp['errors'];
  if (wantErrors is List<Object?>) {
    final want = [
      for (final e in wantErrors.whereType<Map<String, Object?>>())
        '${e['field_key']}|${e['code']}',
    ]..sort();
    final got = [for (final e in errors) '${e.fieldKey}|${e.code}']..sort();
    if (!deepEqual(want, got)) fail('errors', '', want, got);
  }
  final valid = exp['valid'];
  if (valid is bool && valid != errors.isEmpty) {
    fail('valid', '', valid, errors.isEmpty);
  }
}

/// Runs one test case (`{name, context?, lists?, steps?, expect?}`) against
/// [form]. [lists] replaces the case's own `lists`.
TestCaseResult runTestCase(
  Map<String, Object?> form,
  Map<String, Object?> testCase, {
  FormLists? lists,
}) {
  final name = testCase['name'];
  final context = contextFromSnapshot(testCase['context']);
  final optionLists = lists ?? FormLists.fromJson(testCase['lists']);
  final state = <String, Object?>{};
  final failures = <TestFailure>[];

  void evaluateAndCheck(Map<String, Object?> expect, Object step) {
    final resolved = resolveForm(
      form,
      context: context,
      answers: state,
      lists: optionLists,
    );
    final res = validateAnswers(
      form,
      buildAnswers(form, resolved, state),
      context: context,
      lists: optionLists,
    );
    _check(expect, res.resolved ?? resolved, res.errors, step, failures);
  }

  try {
    final steps = testCase['steps'];
    if (steps is List<Object?>) {
      for (var i = 0; i < steps.length; i++) {
        final step = steps[i];
        if (step is! Map<String, Object?>) continue;
        final set = step['set'];
        if (set is Map<String, Object?>) state.addAll(set);
        final unset = step['unset'];
        if (unset is List<Object?>) unset.forEach(state.remove);
        final expect = step['expect'];
        if (expect is Map<String, Object?>) evaluateAndCheck(expect, i);
      }
    }
    final expect = testCase['expect'];
    if (expect is Map<String, Object?>) evaluateAndCheck(expect, 'final');
  } on Exception catch (e) {
    failures.add(TestFailure('final', 'exception', '', null, '$e'));
  }
  return TestCaseResult(name is String ? name : '', failures);
}

/// Runs every case; passes when all of them do.
({bool passed, List<TestCaseResult> results}) runTestCases(
  Map<String, Object?> form,
  List<Map<String, Object?>> cases, {
  FormLists? lists,
}) {
  final results = [
    for (final c in cases) runTestCase(form, c, lists: lists),
  ];
  return (passed: results.every((r) => r.passed), results: results);
}
